TowerScheduler/scheduler.py

```diff
@@ -0,0 +1,442 @@
+        self.logger = get_logger("Sleep State", self.config.log_level)
+
+    @classmethod
+    def get_singleton(cls):
+        if Sleep.singleton is None:
+            Sleep.singleton = Sleep()
+        return Sleep.singleton
+
+    def process(self, state_machine):
+        self.logger.info("Running Sleep process func")
+
+        # write curr index to ens file before calcs
+        with open("active_ensembles.json", "w", encoding="utf-8") as f_out:
+            json_file = {
+                "ensemble_list": Ensemble.list_to_json(state_machine.ens_list),
+                "next_ensemble": state_machine.ens_index
+            }
+            json.dump(json_file, f_out, indent=4)
+
+        now = dt.datetime.now()
+        self.nearest_ens_time = dt.datetime.combine(dt.date.today(),
+                state_machine.ens_list[state_machine.ens_index].start_time)
+
+        self.logger.info("Next ensemble is at %02i:%02i:%02i, and " + \
+                "current time is %02i:%02i:%02i",
+                    self.nearest_ens_time.hour,
+                    self.nearest_ens_time.minute,
+                    self.nearest_ens_time.second,
+                    now.hour, now.minute, now.second)
+
+    def update(self, state_machine):
+        self.logger.info("Running Sleep update func")
+
+        buffer = self.config.wakeup_time + self.config.shutdown_time
+
+        # write configuration to save any updates before sleep
+        self.config.write()
+
+        # recalc current_time vs (current_ensemble time + wakeup + shutdown)
+        now = dt.datetime.now()
+
+        if self.nearest_ens_time.time() < now.time():
+            self.logger.info("Last ensemble finished, index reset; " + \
+                            "next ens is on next day")
+            max_today = dt.datetime.combine(dt.date.today(), dt.time.max)
+            to_nearest = dt.timedelta(hours=self.nearest_ens_time.hour,
+                                    minutes=self.nearest_ens_time.minute,
+                                    seconds=self.nearest_ens_time.second)
+
+            left_today = max_today - now
+            available_sleep_time = (left_today + to_nearest).total_seconds()
+        else:
+            self.logger.info("Current time less than next ens; " + \
+                            "next ens is on same day")
+            available_sleep_time = (self.nearest_ens_time - now).total_seconds()
+
+        # if enough time, call shutdown
+        if available_sleep_time > buffer:
+            to_sleep = available_sleep_time - buffer
+            self.logger.info("calling sleep timer's sleep(%i)", to_sleep)
+            state_machine.sleep_func(to_sleep)
+            time.sleep(1) # yield
+
+            # Python sleep, sleep timer is offline
+            self.logger.error("Sleep timer failed to shut tower down. " + \
+                "Calling Python's time.sleep for %i seconds", to_sleep)
+            time.sleep(to_sleep)
```

Comment:
You're going to be late! You need to recompute `to_sleep` after sleep_func failure.